mysys/crc32/crc32_arm64.c

```diff
@@ -68,250 +74,194 @@ asm(".arch_extension crypto");
+      t0= (uint64_t)vmull_p64(crc0, k1);
+      crc= __crc32cd(crc2, *(const uint64_t *)buffer);
+      crc1= __crc32cd(0, t1);
+      crc^= crc1;
+      crc0= __crc32cd(0, t0);
+      crc^= crc0;
+
+      buffer+= sizeof(uint64_t);
+    }
+
+# else /* HAVE_ARMV8_CRC_CRYPTO_INTRINSICS */
+
+    /*No intrinsics*/
+    __asm__("mov    x16,            #0xf38a         \n\t"
+            "movk   x16,            #0xe417, lsl 16 \n\t"
+            "mov    v1.2d[0],       x16             \n\t"
+            "mov    x16,            #0x8014         \n\t"
+            "movk   x16,            #0x8f15, lsl 16 \n\t"
+            "mov    v0.2d[0],       x16             \n\t"
+            :::"x16");
+
+    while ((length-= 1024) >= 0)
+    {
+      PREF1KL2(buffer, 1024*3);
+      __asm__("crc32cx %w[c0], %w[c], %x[v]\n\t"
+              :[c0]"=r"(crc0):[c]"r"(crc), [v]"r"(*(const uint64_t *)buffer):);
+      crc1= 0;
+      crc2= 0;
+      buffer+= sizeof(uint64_t);
+
+      CRC32C7X3X8(buffer, 0);
+      CRC32C7X3X8(buffer, 1);
+      CRC32C7X3X8(buffer, 2);
+      CRC32C7X3X8(buffer, 3);
+      CRC32C7X3X8(buffer, 4);
+      CRC32C7X3X8(buffer, 5);
+
+      buffer+= 42*3*sizeof(uint64_t);
+      PREF1KL1(buffer, 1024);
+      __asm__("mov            v2.2d[0],       %x[c1]          \n\t"
+              "pmull          v2.1q,          v2.1d,  v0.1d   \n\t"
+              "mov            v3.2d[0],       %x[c0]          \n\t"
+              "pmull          v3.1q,          v3.1d,  v1.1d   \n\t"
+              "crc32cx        %w[c],          %w[c2], %x[v]   \n\t"
+              "mov            %x[c1],         v2.2d[0]        \n\t"
+              "crc32cx        %w[c1],         wzr,    %x[c1]  \n\t"
+              "eor            %w[c],          %w[c],  %w[c1]  \n\t"
+              "mov            %x[c0],         v3.2d[0]        \n\t"
+              "crc32cx        %w[c0],         wzr,    %x[c0]  \n\t"
+              "eor            %w[c],          %w[c],  %w[c0]  \n\t"
+              :[c1]"+r"(crc1), [c0]"+r"(crc0), [c2]"+r"(crc2), [c]"+r"(crc)
+              :[v]"r"(*((const uint64_t *)buffer)));
+      buffer+= sizeof(uint64_t);
+    }
+# endif /* HAVE_ARMV8_CRC_CRYPTO_INTRINSICS */
+
+    /* Done if Input data size is aligned with 1024  */
+    if(!(length+= 1024))
+      return (~crc);
 
 #endif /* HAVE_ARMV8_CRYPTO */
 
-		while ((length -= sizeof(uint64_t)) >= 0) {
-			CRC32CX(crc, *(uint64_t *)buffer);
-			buffer += sizeof(uint64_t);
-		}
-		/* The following is more efficient than the straight loop */
-		if (length & sizeof(uint32_t)) {
-			CRC32CW(crc, *(uint32_t *)buffer);
-			buffer += sizeof(uint32_t);
-		}
-		if (length & sizeof(uint16_t)) {
-			CRC32CH(crc, *(uint16_t *)buffer);
-			buffer += sizeof(uint16_t);
-		}
-		if (length & sizeof(uint8_t))
-			CRC32CB(crc, *buffer);
-
-	} else {
-#ifdef HAVE_ARMV8_CRYPTO
-#ifdef HAVE_ARMV8_CRC_CRYPTO_INTRINSICS
-		const poly64_t k1 = 0xe417f38a;
-		uint64_t t0;
-		while ((length -= 1024) >= 0) {
```

Comment:
Very interesting: the entire `else {` branch was only for the case `buffer=NULL` that should never be needed. I think that this removal must be mentioned in the commit message.